### ISSUE TITLE
Relax definitional equality of `transp`

### ISSUE THREAD

**Issue by howsiyu** [agda/agda#5869]
```agda
open import Cubical.Core.Primitives

data Interval : Type where
  left right : Interval
  path : left ≡ right

module _ {ℓ : Level} (P : I → Type ℓ) where
  foo : Interval → Type ℓ
  foo left = P i0
  foo right = P i1
  foo (path i) = P i

  bar : Interval → Interval → Type ℓ
  bar left left = P i1
  bar left right = P i0
  bar left (path j) = P (~ j)
  bar right _ = P i1
  bar (path i) left = P i1
  bar (path i) right = P i
  bar (path i) (path j) = P (i ∨ ~ j)

  baz : (x y : Interval) → bar x y → foo x
  baz left left z = transp (λ k → P (~ k)) i0 z
  baz left right z = z
  baz left (path j) z = transp (λ k → P (~ (j ∨ k))) j z
  baz right _ z = z
  baz (path i) left z = transp (λ k → P (i ∨ ~ k)) i z
  baz (path i) right z = z
  baz (path i) (path j) z = transp (λ k → P (i ∨ ~ (j ∨ k))) (i ∨ j) z
```

The code above typechecks. Just focusing on the last line, If I understand the requirement of `transp` correctly, this implies that `P (i ∨ ~ (j ∨ k))` is constant wrt `k` when `i ∨ j = i1`. However, this cannot be deduced merely from identities of De Morgan Algebra. Consider the De Morgan Algebra `{0, 1, i, j}` where `~ i = i`, `~ j = j`, `i ∧ j = 0`, and `i ∨ j = 1`. `i ∨ ~ (j ∨ 0) = 1` and `i ∨ ~ (j ∨ 1) = i` are different.

**Comment by andreasabel** [comment]
> ```agda
> baz (path i) (path j) z = transp (λ k → P (i ∨ ~ (j ∨ k))) (i ∨ j) z
> ```

Isn't it that `i ∨ j = i1` implies that either `i = i1` or `j = i1` (1)?  In both cases, `i ∨ ~ (j ∨ k)` is constant in `k`.
Which De Morgan Algebra would refute (1)?

Btw., I think the relevant line in the Agda code base is: https://github.com/agda/agda/blob/505464961f07f0991263708fd8cbf5f7ad12f53f/src/full/Agda/TypeChecking/Primitive/Cubical.hs#L1809

**Comment by howsiyu** [comment]
If we are just using the rules of De Morgan Algebra, `i ∨ j = i1` does not imply `i = i1` or `j = i1`. I have provided a counterexample in the original comment.

**Comment by sattlerc** [comment]
> However, this cannot be deduced merely from identities of De Morgan Algebra.

That's right. But it's not needed.

The implication `i ∨ j = 1 → P constant` holds whenever for all interval context substitutions in which `i ∨ j = 1` holds, we also have `P constant`.

The interval presheaf at some interval context K is the elements of the **free** de Morgan algebra on K. And in a free de Morgan algebra, `i ∨ j = 1` implies `i = 1` or `j = 1`.

**Comment by andreasabel** [comment]
> The interval presheaf at some interval context K is the elements of the **free** deMorgan algebra on K. And in a free deMorgan algebra, `i ∨ j = 1` implies `i = 1` or `j = 1`.

Thanks, @sattlerc.  While we are at it, is it sound to reason in the real interval [0;1] with min, max, and (1-) to derive theorems of a free de Morgan algebra?  In other words, would the real interval be a complete model of de Morgan algebras, or is there a theorem that is true in [0;1] but not in all de Morgan algebras?

**Comment by howsiyu** [comment]
```agda
open import Cubical.Core.Primitives
open import Cubical.Foundations.Prelude using (refl)

data Interval : Type where
  left right : Interval
  path : left ≡ right

module _ {ℓ : Level} (P : I → Type ℓ) where
  foo : Interval → Type ℓ
  foo left = P i0
  foo right = P i1
  foo (path i) = P i

  bar : (l : Interval) → foo l → P i0
  bar left a = a
  bar right b = transp (λ j → P (~ j)) i0 b
  bar (path i) z = transp (λ j → P (i ∧ ~ j)) (~ i) z

  baz : (l : Interval) → foo l → P i1
  baz left a = transp (λ j →  P j) i0 a
  baz right b = b
  baz (path i) z = transp (λ j → P (i ∨ j)) i z

  qux : (l : Interval) (z : foo l) → PathP P (bar l z) (baz l z)
  qux left a i = transp (λ j → P (i ∧ j)) (~ i) a
  qux right b i = transp (λ j → P (i ∨ ~ j)) i b
  qux (path k) z i = transp (λ j → P ((i ∧ k) ∨ (i ∧ j) ∨ (k ∧ ~ j))) ((i ∧ k) ∨ ~ (i ∨ k)) z

  quux : (l : Interval) (z : foo l) → PathP P (bar l z) (baz l z)
  quux left a i = transp (λ j → P (i ∧ j)) (~ i) a
  quux right b i = transp (λ j → P (i ∨ ~ j)) i b
  quux (path k) z i = transp (λ j → P ((i ∧ k) ∨ (i ∧ j) ∨ (k ∧ ~ j))) ((i ∨ ~ k) ∧ (~ i ∨ k)) z

  bla : (l : Interval) (z : foo l) → qux l z ≡ quux l z
  bla left z = refl
  bla right z = refl
  bla (path i) z = {!refl!}
```

In that case, should `qux` and `quux` be definitionally equal? They are the same except for the second argument in `transp`, and the second argument is only used to decide which face of the square to check the first argument is constant.

**Comment by sattlerc** [comment]
> In that case, should `qux` and `quux` be definitionally equal? They are the same except for the second argument in `transp`, and the second argument is only used to decide which face of the square to check the first argument is constant.

No, these are different `transp` instances. The second argument is relevant.

**Comment by howsiyu** [comment]
My point is even though their second arguments are different, they determine the same set of faces in the square.

**Comment by sattlerc** [comment]
They have different reduction behaviour: when the proposition encoded by the second argument is true, the transport is guaranteed to evaluate to the third argument. So they cannot be identified.

**Comment by sattlerc** [comment]
> While we are at it, is it sound to reason in the real interval [0;1] with min, max, and (1-) to derive theorems of a free de Morgan algebra? In other words, would the real interval be a complete model of de Morgan algebras, or is there a theorem that is true in [0;1] but not in all de Morgan algebras?

No. For example, the real interval validates `∀ x, y. x = min(x, y) ∨ x = max(x, y)`, but the free de Morgan algebra on two variables does not.

**Comment by howsiyu** [comment]
But both `(i ∧ k) ∨ ~ (i ∨ k) = 1` and `(i ∨ ~ k) ∧ (~ i ∨ k) = 1` iff `i = 0, j = 0` or `i = 1, j = 1` in a free De Morgan algebra.

**Comment by howsiyu** [comment]
> > While we are at it, is it sound to reason in the real interval [0;1] with min, max, and (1-) to derive theorems of a free de Morgan algebra? In other words, would the real interval be a complete model of de Morgan algebras, or is there a theorem that is true in [0;1] but not in all de Morgan algebras?
> 
> No. For example, the real interval validates `∀ x, y. x = min(x, y) ∨ x = max(x, y)`, but the free de Morgan algebra on two variables does not.

What if we use [0;1] ^ n instead?

**Comment by sattlerc** [comment]
If we substitute `i = 0` in your two interval expressions (depending on `i` and `k`), then:
* `(i ∧ k) ∨ ~ (i ∨ k)` = `~ k`
* `(i ∨ ~ k) ∧ (~ i ∨ k)` = `~ k ∧ k`.

These are different elements of the free de Morgan algebra on the singleton set of `k`. So the two expressions are not equal.

**Comment by sattlerc** [comment]
> But both `(i ∧ k) ∨ ~ (i ∨ k) = 1` and `(i ∨ ~ k) ∧ (~ i ∨ k) = 1` iff `i = 0, j = 0` or `i = 1, j = 1` in a free De Morgan algebra.

But `(i ∧ k) ∨ ~ (i ∨ k)` and `(i ∨ ~ k) ∧ (~ i ∨ k)` are different elements of the free de Morgan algebra on `i` and `k`.

**Comment by howsiyu** [comment]
`(i ∨ ~ k) ∧ (~ i ∨ k)` becomes `~k` if we substitute `i = 0`.

**Comment by howsiyu** [comment]
> But `(i ∧ k) ∨ ~ (i ∨ k)` and `(i ∨ ~ k) ∧ (~ i ∨ k)` are different elements of the free de Morgan algebra on `i` and `k`.

True but it seems like we only care about the condition when they are equal to `1`.

**Comment by sattlerc** [comment]
> True but it seems like we only care about the condition when they are equal to `1`.

Why?

**Comment by sattlerc** [comment]
> What if we use [0;1] ^ n instead?

Probably.

**Comment by howsiyu** [comment]
> > True but it seems like we only care about the condition when they are equal to `1`.
> 
> Why?

In [A CUBICAL TYPE THEORY FOR HIGHER INDUCTIVE TYPES](http://www.cse.chalmers.se/%7Esimonhu/misc/hcomp.pdf), transp is described as
> The generalization of the transport operation from [1] where one can also specify where the given type is known to be constant; on this part the output is equal to the input.

**Comment by andreasabel** [comment]
> But `(i ∧ k) ∨ ~ (i ∨ k)` and `(i ∨ ~ k) ∧ (~ i ∨ k)` are different elements of the free de Morgan algebra on `i` and `k`.

Yes, these are their normal forms (I hacked together a normalizer...):
```
ghci> run 2 "(i ∧ k) ∨ ~ (i ∨ k)"
~ i ∧ ~ k ∨ i ∧ k
ghci> run 2 "(i ∨ ~ k) ∧ (~ i ∨ k)"
~ i ∧ ~ k ∨ ~ i ∧ i ∨ ~ k ∧ k ∨ i ∧ k
```
(Note that conjunction binds stronger than disjunction here.)

@howsiyu : So you are saying we should identify `transp A phi1` and `transp A phi2` already when `phi1 = 1` _iff_ `phi2 = 1`? 
Afaict, atm they are only identified when `phi1 <==> phi2`, via this function: https://github.com/agda/agda/blob/505464961f07f0991263708fd8cbf5f7ad12f53f/src/full/Agda/TypeChecking/Conversion.hs#L1950

**Comment by sattlerc** [comment]
@howsiyu I think I understand now what you mean. You are asking why we take an interval variable `i` as second argument instead of the proposition encoded by `i = 1`.

Indeed, that's how cubical type theory has been defined, with the cofibration classifier (of which the second argument of transport is an element) a strictly univalent universe of propositions.

In the implementation in cubical Agda, the interval sort itself has been taken as the cofibration classifier (with interpretation function sending `i` to `i = 1`). It would have been cleaner to follow the presentation using a separate cofibration classifier.

**Comment by howsiyu** [comment]
> @howsiyu : So you are saying we should identify `transp A phi1` and `transp A phi2` already when `phi1 = 1` _iff_ `phi2 = 1`? Afaict, atm they are only identified when `phi1 <==> phi2`, via this function:

Yes. If we are working in general De Morgan Algebra `phi1 = 1` _iff_ `phi2 = 1` is the same as `phi1 = phi2`. However, if we are working in free De Morgan Algebra `phi1 = 1` _iff_ `phi2 = 1` is weaker than `phi1 = phi2` and I claim the former suffices for our purpose.

**Comment by howsiyu** [comment]
> @howsiyu I think I understand now what you mean. You are asking why we take an interval variable `i` as second argument instead of the proposition encoded by `i = 1`.

Yes.

**Comment by andreasabel** [comment]
> > @howsiyu : So you are saying we should identify `transp A phi1` and `transp A phi2` already when `phi1 = 1` _iff_ `phi2 = 1`? Afaict, atm they are only identified when `phi1 <==> phi2`, via this function:
> 
> Yes. If we are working in general De Morgan Algebra `phi1 = 1` _iff_ `phi2 = 1` is the same as `phi1 = phi2`. However, if we are working in free De Morgan Algebra `phi1 = 1` _iff_ `phi2 = 1` is weaker than `phi1 = phi2` and I claim the former suffices for our purpose.

This would be a relaxation of the definitional equality which could be beneficial.  Most likely, the building blocks to implement this relaxation are already present in the code base and just have to be put together.
Does this only affect `transp` or also other cubical primitives?

**Comment by sattlerc** [comment]
It is preferable to implement the cofibration classifier `Cof` as its own sort, separate from the interval sort `I`. It is true that we have two obvious surjections `I → Cof`, but it is ugly to speak about `Cof` using an equivalence relation on `I`.

Basically, `Cof` is generated by the strict propositions `i = 0` and `i = 1` for `i : I` (only one of them is needed in the presence of negation, but removing negation, or hiding it under an option, may be necessary in the future anyway to retain compatibility with standard homotopy types; so we should write new stuff with that in mind), finite disjunctions, dependent sums, and universal quantification over the interval sort. Definitional equality in `Cof` given by equivalence of propositions.

One has normal forms for cofibrations given by finite disjunctions of finite conjunctions of `i = 0` and `i = 1` (see the cubical type theory papers), so equality is decidable. But reduction to normal forms may introduce exponential blow-up, so we might wish to avoid them.

All cubical primitives except path types would be impacted, but in a uniform way (replacing the interval argument coding a cofibration by a cofibration argument), so this should be more like a refactoring.

@Saizan would know how to do change this (since he made the decision to use `I` in place of `Cof` in the first place). If he is unavailable, I can assist someone who is familiar with the code base.

**Comment by Saizan** [comment]
`(~ r) = 0` is not the same as `Not (r = 0)`, in fact handling Not in Cof would probably be complicated: `Not (i = 0)` does not give you an assignment for `i`.

So in that sense using `I` as argument for the various primitives gives you a little more expressivity, at the cost of more distinctions on that argument. We used to rely on this in some of the reductions for transp and hcomp, not sure anymore.

If we want to get rid of reversals anyway then I guess this is moot, but that's a larger change we should do as we move to a cartesian cubical type theory?

**Comment by sattlerc** [comment]
My point was that `(~ r) = 0` is the same as `r = 1`, so in the presence of negations in the interval, only one of the equalities `i = 0`, `i = 1` is needed. But that seems like a hack to me (that depends on reversals).

I don't think we need negation of cofibrations.

EDIT: I should have used the term _reversal_ instead.

**Comment by Saizan** [comment]
I see, I misunderstood that aside, anyhow my main point was that if you try to refactor `I` to `Cof` you might bump into places where we made use of reversals to "optimize" computations. Maybe this was mostly in the compGlue implementation we dropped, but it could show up elsewhere.
